sycl/source/detail/stream_impl.cpp

```diff
@@ -17,27 +17,55 @@ namespace detail {
 
 stream_impl::stream_impl(size_t BufferSize, size_t MaxStatementSize,
                          handler &CGH)
-    : BufferSize_(BufferSize), MaxStatementSize_(MaxStatementSize),
-      // Allocate additional place for the offset variable and the end of line
-      // symbol. Initialize buffer with zeros, this is needed for two reasons:
-      // 1. We don't need to care about end of line when printing out streamed
-      // data.
-      // 2. Offset is properly initialized.
-      Data(BufferSize + OffsetSize + 1, 0),
-      Buf(Data.data(), range<1>(BufferSize + OffsetSize + 1),
-          {property::buffer::use_host_ptr()}),
-
-      FlushBuf(range<1>(MaxStatementSize)) {}
+    : BufferSize_(BufferSize), MaxStatementSize_(MaxStatementSize) {
+  // We need to store stream buffers in the scheduler because they need to be
+  // alive after submitting the kernel. They cannot be stored in the stream
+  // object because it causes loop dependency between objects and results in
+  // memory leak.
+  // Allocate additional place in the stream buffer for the offset variable and
+  // the end of line symbol.
+  detail::Scheduler::getInstance().allocateStreamBuffers(
+      this, BufferSize + OffsetSize + 1 /* size of the stream buffer */,
+      MaxStatementSize /* size of the flush buffer */);
+}
+
+// Method to provide an access to the global stream buffer
+GlobalBufAccessorT stream_impl::accessGlobalBuf(handler &CGH) {
+  return detail::Scheduler::getInstance()
+      .StreamBuffersPool.find(this)
+      ->second.Buf.get_access<cl::sycl::access::mode::read_write>(
+          CGH, range<1>(BufferSize_), id<1>(OffsetSize));
+}
 
+// Method to provide an accessor to the global flush buffer
+GlobalBufAccessorT stream_impl::accessGlobalFlushBuf(handler &CGH) {
+  return detail::Scheduler::getInstance()
+      .StreamBuffersPool.find(this)
+      ->second.FlushBuf.get_access<cl::sycl::access::mode::read_write>(
+          CGH, range<1>(MaxStatementSize_), id<1>(0));
+}
+
+// Method to provide an atomic access to the offset in the global stream
+// buffer and offset in the flush buffer
+GlobalOffsetAccessorT stream_impl::accessGlobalOffset(handler &CGH) {
+  auto OffsetSubBuf = buffer<char, 1>(
+      detail::Scheduler::getInstance().StreamBuffersPool.find(this)->second.Buf,
+      id<1>(0), range<1>(OffsetSize));
+  auto ReinterpretedBuf = OffsetSubBuf.reinterpret<unsigned, 1>(range<1>(2));
+  return ReinterpretedBuf.get_access<cl::sycl::access::mode::atomic>(
+      CGH, range<1>(2), id<1>(0));
+}
```

Comment:
Are these functions guarded to be available at host only?

Reply:
Yes, correct